refactor: submitted transactions

fbwoolf opened the pull request:
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/stacks-wallet-web/actions/runs/2740309397).<!-- Sticky Header Marker -->

This PR removes submitted transactions from local storage which allows us to remove lots of code, including `atomWithStorage` and other deprecated atoms. This proposal gets us to one atom `accountSubmittedTransactionsState`. I need to make another pass thru to clean up the actual components further bc I mostly just moved/renamed/separated them, but wanted to get your feedback on this @kyranjamie at this point. I chose to go with the naming `SubmittedTransaction` rather than opting for the word 'local' which seemed confusing to me at times.

*I also left a comment to add all possible error reasons, which I would like to include here for better error handling.

EDIT: Possible move to redux here would be easy now.

cc/ @kyranjamie @fbwoolf

kyranjamie commented on src/app/features/activity-list/components/submitted-transactions-list.tsx:
```diff
@@ -0,0 +1,40 @@
+  txs: SubmittedTransaction[];
+}
+export const SubmittedTransactionsList = ({ txs }: SubmittedTransactionsListProps) => {
+  const cleanup = useCleanupSubmittedTxsCallback();
+
+  useEffect(() => cleanup(), [cleanup]);
```
This part of the implementation I find unusual in that, what does the component life cycle have to do with the submitted transaction state?

It'd be better if the two were not coupled. Changing/moving around components, a different design/ux of the tx list, could alter when this is called.

fbwoolf replied:
Yeah, agree, I wanted to get stuff working but wasn't clear on why this is here so I'll rethink it.

kyranjamie commented on src/app/store/accounts/submitted-transactions.ts:
```diff
@@ -0,0 +1,8 @@
+  txid: string;
+}
+
+export const accountSubmittedTransactionsState = atom<SubmittedTransaction[]>([]);
```
I think it does make sense to use an object keyed by txid, rather than an array. This avoids the need to filter duplicates/check if a tx of same id exists. If you broadcast twice and it were to have same txid, just update it (no need for additional handling)

fbwoolf replied:
Ok, do you think we need to keep the timestamp for any reason?

kyranjamie replied:
As long as we're not persisting it long term, I don't reckon so

fbwoolf replied:
Thinking about this, I guess I don't understand needing to use an object. We are NOT doing checks to filter duplicates in the array itself ...we are checking account transactions (confirmed/mempool) to move it in/out of the submitted state if needed. Am I understanding your comment correctly?

kyranjamie commented on src/app/features/activity-list/components/submitted-transactions.utils.ts:
```diff
@@ -0,0 +1,64 @@
+        withTicker: false,
+      })}`;
+    case PayloadType.ContractCall:
+    // TODO: Implement?
+    // return transaction.payload.functionName.content === 'transfer'
+    //   ? cvToValue(transaction.payload.functionArgs[0])
+    //   : null;
```
would avoid and keep simple (any method could be called transfer and doesn't necessarily mean transferring assets

fbwoolf replied:
Ok, yeah, this comment was here so didn't want to just remove w/out checking. I'll remove it.

kyranjamie commented on src/app/store/accounts/submitted-transactions.hooks.ts:
```diff
@@ -0,0 +1,58 @@
+  );
+}
+
+export function useAccountSubmittedStacksTransactions() {
+  const [submittedTxs] = useAccountSubmittedTransactionsState();
+  return submittedTxs.map(tx => deserializeTransaction(tx.rawTx));
+}
```
we should memoize these hooks. hooks returning new obj reference can be hard to track down when they cause render loops. also imagine deserializetx to be a fairly expensive operation that's happening on every render where this is used

kyranjamie reviewed (commented):
Great start @fbwoolf. 

Imo it's fine to not persist submitted txs, the happy path is straight to mempool anyway, and we've had so many problems with stuck submitted txs. 

We're good to use redux here too. You have to implement crud with the atoms, this'd better be handled by `createEntityAdapter`

fbwoolf commented on src/app/common/hooks/use-submit-stx-transaction.ts:
```diff
@@ -46,10 +45,9 @@ export function useSubmitTransactionCallback({ loadingKey }: UseSubmitTransactio
             const txid = `0x${response.txid}`;
-            if (!externalTxid.includes(txid)) {
-              await setLocalTxs({
+            if (!accountTxIds.includes(txid)) {
```
This is checking to make sure it isn't already a pending tx, correct? I think we are checking to make sure it needs to go into the submitted state.

kyranjamie commented on src/app/store/transactions/transaction.hooks.ts:
```diff
@@ -188,20 +190,18 @@ export function useSoftwareWalletTransactionBroadcast() {
-                rawTx: result.txRaw,
-                timestamp: todaysIsoDate(),
-              },
+          if (typeof result.txId === 'string' && !accountTxIds.includes(result.txId)) {
```
To your other comment on using an object, this check wouldn't be necessary. Using an object prevents duplicates as you can only ever have one value assigned to a given key.

It's not the end of the world to use an array, if you like, but there are merits to using maps/dictionaries. When using createEntityAdapter, it'll be an object. It's a common approach to take when ensuring state is normalised, see https://redux.js.org/usage/structuring-reducers/normalizing-state-shape

kyranjamie replied:
Sure, sorry it is indeed unrelated to object discussion. 

Though, not sure why it's needed? 🤔  In what circumstance would a pending tx appear before it's submitted?

fbwoolf replied:
Not quite sure, but it was a check used in the send form so added here. Maybe just a safety check to make sure it hasn't already become pending. Maybe it could happen that fast?

kyranjamie replied:
I see. I commend your anti-race condition approach. I don't think there's any chance of this happening, but I suppose it doesn't hurt.

fbwoolf replied:
Haha, I can remove it and see if we run into any issues during testing.

fbwoolf commented:
I addressed the basic code comments here, now I'm going to refactor to redux using `createEntityAdapter` as a separate commit.

kyranjamie commented on src/app/features/activity-list/components/submitted-transactions-list.tsx:
```diff
@@ -0,0 +1,29 @@
+const SubmittedTransactionsListItem = ({ txid }: { txid: string }) => {
+  const submittedTx = useSubmittedStacksTransaction(txid);
+  if (submittedTx) return <SubmittedTransactionsItem transaction={submittedTx} txid={txid} />;
+  return <>Loading...</>;
```
I imagine we want to return null? A request to render a submitted tx that doesn't existing it state should display nothing

fbwoolf replied:
Yep, that was here and just missed it, thx. I'll remove it.

fbwoolf commented on src/app/features/activity-list/components/submitted-transactions-list.tsx:
```diff
@@ -0,0 +1,29 @@
+      </Text>
+      <Stack mt="base-loose" spacing="loose">
+        {txs.map(tx => (
+          <SubmittedTransactionsListItem key={tx.txid} txid={tx.txid} />
```
Good idea, thx.

fbwoolf commented on src/app/store/accounts/submitted-transactions.hooks.ts:
```diff
@@ -0,0 +1,66 @@
+  const accountTxIds = useCurrentAccountTxIds();
+  const removeSubmittedTransaction = useRemoveSubmittedTransactionCallback();
+
+  return useEffect(() => {
+    submittedTxs.map(tx => {
+      if (accountTxIds.includes(safelyFormatHexTxid(tx.txid))) removeSubmittedTransaction(tx.txid);
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [accountTxIds, submittedTxs]);
```
Yep, thx, it was a useCallback and missed removing.

fbwoolf commented:
@kyranjamie this has been refactored to use redux and is ready for review again, thx!

fbwoolf commented on src/app/common/get-error-message.ts:
```diff
@@ -3,20 +3,59 @@ import { TxBroadcastResultRejected } from '@stacks/transactions';
 export function getErrorMessage(
```
Not 100% sure my return messages are what we want to show here, so maybe let me know changes and/or if we should just show some as the default message. I did make sure they are all accounted from here now.

In addition to the api, ref: https://github.com/stacks-network/stacks-blockchain/blob/master/src/chainstate/stacks/db/blocks.rs#L259

kyranjamie commented on src/app/common/hooks/use-submit-stx-transaction.ts:
```diff
@@ -45,14 +44,12 @@ export function useSubmitTransactionCallback({ loadingKey }: UseSubmitTransactio
+              submittedTransactionsActions.transactionAdded({
                 rawTx: transaction.serialize().toString('hex'),
-                timestamp: todaysIsoDate(),
-                txid,
-              });
-            }
+                txId: safelyFormatHexTxid(response.txid),
+              })
```
I'd create a hook that exposes a method calling this dispatch (and other actions from this slice)

kyranjamie commented on src/app/store/submitted-transactions/submitted-transactions.slice.ts:
```diff
@@ -0,0 +1,22 @@
+import { createEntityAdapter, createSlice, EntityId, PayloadAction } from '@reduxjs/toolkit';
+
+export type SubmittedTransaction = { rawTx: string; txId: string };
```
```suggestion
export interface SubmittedTransaction { rawTx: string; txId: string };
```

unless using the features of type, we should always default to interface, so that use of type communicates something more complex/a composition of other types.

This could also be reused in the component that takes these same props

fbwoolf replied:
Ok, I was following the `createEntityAdapter` docs and they use a type in the example.

`type Book = { bookId: string; title: string }`

kyranjamie replied:
The reality is that either can be used, and it comes down to some minor differences in behaviour and personal pref

fbwoolf replied:
Ok, thanks, I haven't been super clear on the difference tbh so was reading up yest on it before I used type from the example. What I found wasn't super helpful.

kyranjamie commented on src/app/features/activity-list/components/submitted-transactions-list.layout.tsx:
```diff
@@ -0,0 +1,27 @@
+import { SubmittedTransactionsItem } from '@app/features/activity-list/components/submitted-transactions-item';
+import { deserializeTransaction } from '@stacks/transactions';
+
+export const SubmittedTransactionsListItem = ({ rawTx, txId }: { rawTx: string; txId: string }) => {
```
```suggestion
type SubmittedTransactionsListItemProps = SubmittedTransaction;
export const SubmittedTransactionsListItem = ({ rawTx, txId }: SubmittedTransactionsListItemProps) => {
```

kyranjamie replied:
Using type here as, for consistency, renaming the interface for use as a components props

kyranjamie commented on src/app/features/activity-list/hooks/use-current-submitted-transactions.ts:
```diff
@@ -0,0 +1,24 @@
+import { safelyFormatHexTxid } from '@app/common/utils/safe-handle-txid';
+import { useAppDispatch } from '@app/store';
+
+export function useCurrentSubmittedTransactions() {
+  const submittedTransactions = useSubmittedTransactions();
+  const accountTxIds = useCurrentAccountTxIds();
+  const dispatch = useAppDispatch();
+
+  useEffect(() => {
+    Object.values(submittedTransactions).map(tx => {
+      if (!tx) return;
+      if (accountTxIds.includes(safelyFormatHexTxid(tx.txId)))
+        dispatch(submittedTransactionsActions.transactionRemoved(tx.txId));
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [accountTxIds, submittedTransactions]);
+
+  return submittedTransactions;
+}
```
Really curious to see how we could kill this hook/reimplement another way

fbwoolf replied:
I'll try to move this today ...seems like it just needs to update when pending txs update.

kyranjamie replied:
Also, can you check often this logic executes? Looking at `useCurrentAccountTxIds`

https://github.com/hirosystems/stacks-wallet-web/blob/3a890d665a70241a9593859ea889c2c0c897290d/src/app/query/transactions/transaction.hooks.ts#L10-L13

it calls `.map()`, which on every render, will return a new array instance. Then, on every render, `useEffect` here will detect it as a new array, triggering this logic

kyranjamie commented on src/app/store/submitted-transactions/submitted-transactions.selectors.ts:
```diff
@@ -0,0 +1,15 @@
+
+const selectSubmittedTransactionsSlice = (state: RootState) => state.submittedTransactions;
+
+export const selectSubmittedTransactions = createSelector(
+  selectSubmittedTransactionsSlice,
+  state => state.entities
+);
```
Can use submitedTxAdapter to generate this automatically

fbwoolf replied:
Do you know why you did it this way for the key selector?

kyranjamie replied:
I do not, this could also use `keyAdapter.getSelectors().selectById`

fbwoolf replied:
What is best in your opinion, to use `selectEntities` and use the dictionary (which is how the code is now), or `selectAll` and use an array of entities.

kyranjamie replied:
Afaict, you're only ever consuming it as an array, so better to export as an array `selectAll`. 

If, at some point, we need to find a specific one by id, I'd create another higher order selector that takes an id param (sooner than using `array.find`)

kyranjamie commented on src/app/store/submitted-transactions/submitted-transactions.slice.ts:
```diff
@@ -0,0 +1,22 @@
+  name: 'submittedTransactions',
+  initialState: initialSubmittedTransactionsState,
+  reducers: {
+    transactionAdded(state, action: PayloadAction<SubmittedTransaction>) {
+      submittedTransactionsAdapter.addOne(state, action.payload);
+    },
+    transactionRemoved(state, action: PayloadAction<EntityId>) {
+      submittedTransactionsAdapter.removeOne(state, action.payload);
+    },
```
This action names describe the result we want from an action or event. Per https://redux.js.org/style-guide/#model-actions-as-events-not-setters better to describe the event itself.

`newTransactionSubmitted` or `transactionBroadcastSuccessfully`, and `transactionEnteredMempool`?

fbwoolf replied:
Ok, again, was referring to the example in docs.

`bookAdded: booksAdapter.addOne`

kyranjamie replied:
Sure, I see your thinking. I guess in some simple cases, the event is pretty much is the setter: If I "add a book to my collection" that maps 1:1 with adding a book to the state.

Here adding a submitted transaction to the result we want to happen after successfully broadcasting a transaction. If we were to have a broadcast tx action, we might have even reused that.

fbwoolf commented:
@kyranjamie can you circle back and check where I moved the logic from the removed useEffect hook? Let me know what you think ...I moved it to the mempool transactions query `onSuccess`.

kyranjamie commented on src/app/store/submitted-transactions/submitted-transactions.hooks.ts:
```diff
@@ -0,0 +1,26 @@
+    dispatch(submittedTransactionsActions.transactionReplacedByFee(txId));
+  }
+
+  return {
+    dispatchNewTransactionSubmitted,
+    dispatchTransactionEnteredMempool,
+    dispatchTransactionReplacedByFee,
+  };
```
I would omit the dispatch from the names, this is a redux level concept we don't need to expose

Also, I wonder if this could cause a render loop/unnecessary rerender issue. If I'm not mistaken, on each render, these functions are redefined with a new reference, so in `useSubmitTransactionCallback` where you add one of these to the dep array, it'll recreate that `useCallback` on every render.

I'd define a single useMemo and define these within that, similar to 

https://github.com/hirosystems/stacks-wallet-web/blob/3e4210ce4849a19c145489070e12bad66f9e1259/src/app/features/ledger/hooks/use-ledger-navigate.ts#L7-L87

kyranjamie commented on src/app/store/submitted-transactions/submitted-transactions.selectors.ts:
```diff
@@ -0,0 +1,11 @@
+);
+
+export function useSubmittedTransactions() {
+  return submittedTransactionsSelectors.selectAll(store.getState());
```
```suggestion
  return useSelector(submittedTransactionsSelectors.selectAll);
```

Should use `useSelector` to read from store. 

Selectors "subscribe" to the store, so that future updates trigger renders. Here, it's different in that we do a one-time read from the store at render-time and return the value. It probably works, as renders happen often, but is subtly different and could be v confusing if, say, a render wasn't triggered.

kyranjamie reviewed (approved):
Great work Fara
